pkg/services/ngalert/schedule/schedule.go

```diff
@@ -297,10 +452,19 @@ func (sch *schedule) ruleRoutine(grafanaCtx context.Context, key models.AlertRul
 				sch.saveAlertStates(processedStates)
 				alerts := FromAlertStateToPostableAlerts(sch.log, processedStates, sch.stateManager, sch.appURL)
 				sch.log.Debug("sending alerts to notifier", "count", len(alerts.PostableAlerts), "alerts", alerts.PostableAlerts)
-				err = sch.sendAlerts(alerts)
+				err = sch.notifier.PutAlerts(alerts)
 				if err != nil {
 					sch.log.Error("failed to put alerts in the notifier", "count", len(alerts.PostableAlerts), "err", err)
 				}
+
+				// Send alerts to external Alertmanager(s) if we have a sender for this organization.
+				sch.sendersMtx.Lock()
+				defer sch.sendersMtx.Unlock()
+				s, ok := sch.senders[alertRule.OrgID]
+				if ok {
+					s.SendAlerts(alerts)
+				}
```

Comment:
2 things:
1. More than 1 rule can get to this step at the same time, and I think it wont be a rare case. So how about using a RW mutex and acquire a read lock here?
2. Doing it in sync gives the control to the external alertmanager for how long we are stuck here and there could be some unexpected delays here. I think that can make some rule evaluations get delayed if they are high frequency? So maybe do it async with a way to make sure that we don't leak goroutines when we are closing this routine.

Reply:
I've done 1. I need to think/investigate a bit more about 2 given the notifiers uses a queue but I can't tell straight away if it'll block or not https://github.com/prometheus/prometheus/blob/main/notifier/notifier.go#L385

Reply:
Right, looks like it wont block, so 2 is not an issue